// @atomist/rug/test/ScenarioWorld.ts
import { Project } from "../model/Project";

/**
 * Superinterface for all worlds: Isolated contexts
 * available during scenario execution
 */
export interface ScenarioWorld {

    /**
     * Get the value of the given key in the scenario, or null
     */
    get(key: string): any;

    /**
     * Bind the value to the given key in the scenario
     */
    put(key: string, what: any): void;

    /**
     * Clear the value of the given key in the scenario
     */
    clear(key: string): void;

    /**
     * Abort execution of the current scenario. This will cause failure.
     */
    abort(): void;

    /**
     * Was execution of the current scenario aborted?
     */
    aborted(): boolean;

    /**
     * Did the last operation fail due to invalid parameters?
     * Otherwise null
     */
    invalidParameters(): any;

    /**
     * Clone the given repo and provide a reference to the project
     */
    cloneRepo(cloneInfo: CloneInfo): Project;

}

/**
 * Identifies a repo
 */
export interface RepoId {

    branch: string;

    sha: string;

    owner: string;

    name: string;

}

/**
 * Information necessary to clone a repo
 */
export class CloneInfo implements RepoId {

    public branch = "master";

    public sha: string;

    public depth: number = 10;

    public remote: string;

    constructor(public owner: string, public name: string) { }

    public withBranch(branch: string): CloneInfo {
        this.branch = branch;
        return this;
    }

    public withSha(sha: string): CloneInfo {
        this.sha = sha;
        return this;
    }

    public withDepth(depth: number): CloneInfo {
        this.depth = depth;
        return this;
    }

    public withRemote(remote: string): CloneInfo {
        this.remote = remote;
        return this;
    }
}
