// @atomist/rug/tree/DecoratingPathExpressionEngine.ts
import {PathExpressionEngine,TreeNode} from "./PathExpression"
import {TransformingPathExpressionEngine} from "./TransformingPathExpressionEngine"
import {TextTreeNodeOps} from "../ast/TextTreeNodeOps"

/**
 * Tries to find an "Ops" class for the given node, and mix in
 * its methods with the node type methods. If an ops class can't
 * be found, mix in TextTreeNodeOps.
 */
export abstract class DecoratingPathExpressionEngine extends TransformingPathExpressionEngine {

    constructor(delegate: PathExpressionEngine) {
        super(delegate, n => {
            // Also need to parameterize module
            let ops = null
            try {
                ops = this.decoratorFor(n)
            }
            catch (e) {
                // Not an error: We didn't find a specific decorator class.
                // Fall back to TextTreeNodeOps, which adds convenience operations.
                //console.log(`Can't find specific decorator for node named ${n.nodeName()}: Expected class name=${this.decoratorClassName(n)}`)
                ops = new TextTreeNodeOps(n, delegate)
            }
            let combined = this.unify(n, ops)
            //console.log(`ops=${ops}, combined=${combined}`)
            return combined
        })
    }

    /**
     * Add all functions from right to left.
     * Also copies state, which is needed for methods to work.
     */
    private unify<T, U>(base: T, enricher: U): T & U {
        let monkeyableBase = base as any
        for (let id in enricher) {
            let fun = (<any>enricher)[id]
            monkeyableBase[id] = fun
            //console.log(`Adding function ${id} from right to left: ${fun} from source ${enricher}`)
        }
        return monkeyableBase as T & U;
    }

    /**
     * Template method subclasses can use to find the decorator for this node.
     * Implementations will typically use the decoratorClassName method.
     * Implementations don't need to catch the error if a class cannot
     * be instantiated: It will be caught in this class.
     */
    protected abstract decoratorFor(n: TreeNode): any

    /**
     * Convenience method returning the conventional decorator
     * class name for the node name. Simply adds "Ops" suffix.
     */
    protected decoratorClassName(n: TreeNode) {
        return n.nodeName().charAt(0).toUpperCase() + n.nodeName().substr(1) + "Ops"
    }

}
