// @atomist/rug/tree/PathExpression.ts

/**
 *  Returned when a PathExpression is evaluated
*/
interface Match<R,N> {

  root(): R

  matches(): N[]
}

/**
 * Object containing a path expression. Facilitates reuse.
 */
class PathExpression<R,N> {

  constructor(public expression: string) {}

}

/**
 * Information about position within a file.
 */
interface FormatInfo {

    start(): PointFormatInfo

    end(): PointFormatInfo
}

interface PointFormatInfo {

   offset(): number
   lineNumberFrom1(): number
   columnNumberFrom1(): number
   indentDepth(): number
   indent(): String
}


/**
 * Operations common to all tree nodes.
 */
interface TreeNode {

  nodeName(): string

  nodeTags(): string[]

  children(): TreeNode[]

}

interface ParentAwareTreeNode extends TreeNode {

 /**
  * Parent node. May be null, depending on the parent node.
  * AST-backed nodes will always have a non-null parent.
  */
  parent(): TreeNode

}

/**
 * Additional operations on text based tree nodes.
 */
interface TextTreeNode extends ParentAwareTreeNode {

  value(): string

  update(newValue: string): void

  /**
   * Information about the position of this node in the current file, if available.
   */
  formatInfo(): FormatInfo

}

interface TypeProvider extends DynamicType {

    typeName: string

    find(context: TreeNode): TreeNode[]

}

/*
 * Interface to execute path expressions and return Match objects.
 */
interface PathExpressionEngine {

/**
 * Add a dynamic type. Can be used as a fluent API.
 */
  addType(dt: DynamicType): PathExpressionEngine

  evaluate<R extends TreeNode,N extends TreeNode>(root: R, expr: PathExpression<R,N>): Match<R,N>

  /**
  * Execute the given function on the nodes returned by the given path expression
  */
  with<N extends TreeNode>(root: TreeNode, expr: string,
            f: (n: N) => void): void

/**
 * Return a single match. Throw an exception otherwise.
 */
  scalar<R extends TreeNode,N extends TreeNode>(root: R, expr: PathExpression<R,N>): N

  /**
   * Convenience method to eval scalar from string
   */
  scalarStr<R extends TreeNode,N extends TreeNode>(root: R, expr: string): N

 /**
 * Cast the present node to the given type
 */
  as<N extends TreeNode>(root, name: string): N

  // Find the children of the current node of this time
  children<N extends TreeNode>(root, name: string): N[]

}

/**
 * Tag interface for dynamic types such as microgrammars
 */
interface DynamicType {

}

/**
 * Dynamic type for a microgrammar. Pass to PathExpression.addType method
 */
class Microgrammar implements DynamicType {

      constructor(public name: string, public grammar: string, public submatchers: any = {}) {}

}

export {Match}
export {PathExpression}
export {PathExpressionEngine}
export {TreeNode}
export {ParentAwareTreeNode}
export {TextTreeNode}
export {FormatInfo}
export {PointFormatInfo}
export {TypeProvider}
export {DynamicType}
export {Microgrammar}
